### DISCUSSION TITLE
Consider introducing SceneRunner Unit Tests

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [SakulFlee/Godot-WebRTC-Match-Maker#169] [Ideas]
https://github.com/MikeSchulze/gdUnit4Mono/issues/35#issuecomment-1953213134

By introducing unit tests that check the camera, it is possible to understand why certain cameras work and not others.

https://github.com/Pieeer1/GodotVoipNet/tree/main/Tests

**Comment by SakulFlee** [comment]
Hi @GeorgeS2019! Sorry for the late reply.
I somehow did miss this notification.

I've been thinking about integrating unit tests, but last time I checked, all current solutions did lack features I was requiring, let alone most solutions being for GDScript only.
I'll check out gdUnit4Mono, might be worth it!

However, the current example app should already print a lot of information to the console.
If there are still issues with getting the camera to work, please share logs of when the example is run, so I can investigate :)
Might have to do with why some virtual cameras are also not showing up.

**Reply by GeorgeS2019** [reply]
>If there are still issues with getting the camera to work, please share logs of when the example is run, so I can investigate :)

I think it is a matter of isolating the steps leading to activation of the camera.

Unit tests certainly will help in isolating the problem

I need to find time to get there.  The original code involving in the camera certainly work. Just need to figure out how the initialization is different when implemented for Godot.

**Reply by GeorgeS2019** [reply]
@Question

Have you thought of jitsi for godot. Now we have Jitsi for Unity, depending on WebRTC for unity.

**Reply by SakulFlee** [reply]
> >If there are still issues with getting the camera to work, please share logs of when the example is run, so I can investigate :)
> 
> I think it is a matter of isolating the steps leading to activation of the camera.
> 
> Unit tests certainly will help in isolating the problem
> 
> I need to find time to get there.  The original code involving in the camera certainly work. Just need to figure out how the initialization is different when implemented for Godot.
> 
> 

I doubt that unit tests will isolate this better than the existing examples. Just attach a debugger and see where it fails, then share logs with me and I can tell you more :)

**Reply by SakulFlee** [reply]
> @Question
> 
> Have you thought of jitsi for godot. Now we have Jitsi for Unity, depending on WebRTC for unity.

Am not familiar with Jitsi, but it seems to be a website communication solution ... I doubt they would like someone making an unofficial 3rd party client for their service and using their website service would require CEF (or similar) and is totally unrelated and decoupled from this project. The WebRTC connection would solely be handled by CEF (or similar) then.

There is [gdcef](https://github.com/Lecrapouille/gdcef) that does exactly that.
Just pipe in the Jitsi website and you are good to go :)
But keep in mind that this would be running separately from this projects WebRTC sockets. So you would end up with 2, or more, sockets.

**Reply by GeorgeS2019** [reply]
> There is [gdcef](https://github.com/Lecrapouille/gdcef) that does exactly that.
## [Not able to get jitsi-meet works in the 2D demo](https://github.com/Lecrapouille/gdcef/issues/49)

**Reply by GeorgeS2019** [reply]
@SakulFlee 
> I'll check out gdUnit4Mono, might be worth it!

I am requesting finePointCGI to use VS2022 Godot4 TestAdapter.  
Have you ever consider writing integration tests, as those available in SipSoceray for this project. This could help new adopters to this project to understand your codes.

Each questions and issues raised could lead to an integration tests, considering the claim you have of the lacking status of Godot4

https://github.com/finepointcgi/WebRTC-with-Godot-Tutorial

**Reply by SakulFlee** [reply]
> > There is [gdcef](https://github.com/Lecrapouille/gdcef) that does exactly that.
> ## [Not able to get jitsi-meet works in the 2D demo](https://github.com/Lecrapouille/gdcef/issues/49)

I see. Still. Jitsi is a 3rd party company that probably wouldn't like people making an unofficial app for their services.
I assume the unity package is a partnership/cooperation between them and not someone just doing making an unofficial package.

That being said, like I said above, different WebRTC protocols are incompatible with each other as they require different features, formats, protocols, etc., to work.
It wouldn't be impossible to figure out what Jitsi's protocol is, but that's out of scope and unofficial at best.
Let alone that many features from Jitsi wouldn't be supported either.

**Reply by SakulFlee** [reply]
> @SakulFlee 
> > I'll check out gdUnit4Mono, might be worth it!
> 
> I am requesting finePointCGI to use VS2022 Godot4 TestAdapter.  
> Have you ever consider writing integration tests, as those available in SipSoceray for this project. This could help new adopters to this project to understand your codes.
> 
> Each questions and issues raised could lead to an integration tests, considering the claim you have of the lacking status of Godot4
> 
> https://github.com/finepointcgi/WebRTC-with-Godot-Tutorial

I still don't see how tests would help people. Rather read the documentation and look at the code examples ...

The existing demo project with all examples practically **are** integration tests. Just without a test adapter behind them and needing to run manually.

That being said, I am also quiet unsure how I can even setup a test environment that:

- boots up a TURN server
  - locally
  - remotely
  - behind a restrictions 
- start at least two WebRTC sessions
  - ideally one on a remote too
- and especially for the video demo: 
  - how to simulate a camera
  - how to simulate a microphone
  - how to capture that data and analysis it

Like yes, I can test parameters being added and processed and such but there really isn't a need for it. And everything else seems overly complicated, especially doing it automatically.
Thus, I really don't see the need for tests like this. Especially since they **already exist** in form of the examples.

If you want them and you can figure those problems out, please implement them and open a PR :)

**Reply by GeorgeS2019** [reply]
I started the TURN server locally. I run another MatchMakerClient.exe app.  I start the unit test, defining it to stay active e.g. 15 seconds. Then I turn on the camera of the Client.exe app. Immediately I turn on the camera of the unit test app to debug under various conditions what need to be improved of the existing implementation

The existing implementation is TOO wired up for ONE perfect condition. If the condition changes slightly, the app fail to run. 

By changing the code, I start to understand the code better, small block by small block at a time.

In this way, the tests help me to learn FAR EFFECTIVE than just having a demo that needs to run in a perfect condition and sequence of events

**Reply by SakulFlee** [reply]
The provided demos/examples serve two purposes:

1. Showcase an _example_ implementation of what a feature like that could look like. It's not necessarily the best and only implementation, just _one possible_ implementation. Thus, naturally, it is "one perfect condition". Though, every example is made to be as simple to understand as possible and as universally as possible. The Video/Audio demo you keep referring to is simply the most extensive one, as it requires a lot of extra work to get stuff going. It's not as simple as say RPC or sending a single message back and forth.
2. Provide a, sort-of, integration test. Without a test framework, yes. My main reason for this is that I commonly feel test-driven development is a big waste of time, for as long as there are better ways of ensuring the product works. Furthermore, if we end up adding tests I want them to be actually fully testing everything, not just a certain part. Especially remote connections will become a problem as we would need to run a TURN server _somewhere_, possibly _publicly_, and somehow pay for that. This project currently is running on zero budget and my free time.

So in some way, you already got your tests there. Integration tests though, not unit tests.
However, I also don't really see the need to add unit tests for ... examples.
You'd be adding a test for a "test" (example). Wouldn't you also need to add a test for your test that tests the examples then? (And so on ...)
I simply don't see the need here.

---

The actual plugin, i.e. "WebRTC (SIPSorcery)" and "Match Maker", are fully untested at the moment, and you could argue tests could be useful here.

However, the WebRTC plugin I provide is basically just a wrapper around SIPSorcery's libraries, which are tested already. I only add some code to expose e.g. events/signals/listeners to Godot and add a way of easily making changes to settings. We could unit test those changes, but honestly I don't see that paying off really. Apart from you wanting to do testing for learning the code ... this wouldn't help you understand the code, as it's literally just settings and signals. If you want to understand the fundamentals, then check out the tests from SIPSorcery! :)
No need to double-check what they already tested for ...
(Same goes for FlashCap and such)

Now, for "Match Maker" I do kinda see the argument.
Though, the current "Match Maker" is an incredibly simple solution and could be potentially exchanged for a simple text file on someone's device. The most important part is how it handles requests and replies ... which the examples/integration tests already pretty much check for. Apart from, you'd also need to run the server somewhere, individually test each request, possibly _with_ using the Godot plugin, which I don't know if that's possible to do in a CI.
_I rather stick to manual QA as it's nothing critical, nor important, can easily be verified and doesn't require any super special circumstances._

That all aside, I also don't see the need for studying the code this deeply.
Like I said, Match Maker is incredibly simple and already well documented.
The Godot plugins are mostly just wrappers to existing libraries.
The demos/examples are already integration tests (_in theory_) and just showcase one of an infinite amount of possible implementations. Mine, by far, isn't the best and will probably never be the best.

From my view, there isn't really much to learn or study here.
You basically have two ways:

## Option 1: With Match Maker

If you want to use the Match Maker, then you only really need to know about a handful of things.
Mainly:

- For the Match Maker server:
  - Compiling & Running Match Maker
  - Configuration file
- For Godot:
  - Compiling & Enabling the Godot plugins (WebRTC & Match Maker)
  - `MatchMaker` node and the listener methods:
    - `MatchMaker::OnMessageString` _OR_ `MatchMaker::OnMessageRaw`

That's all you need. Everything else it automatically done for you.
No further work or studying required. Implementations of using this (e.g. examples) are dependent on the end-user.

## Option 2: Without Match Maker

If you don't want to use the Match Maker, you can do the WebRTC stuff manually.
Therefore, at best, you start studying how WebRTC works.
There is a very good website [here](https://webrtc.github.io/webrtc-org/start/), explaining basically everything you need to know.

The provided `WebRTCPeer` and `WebRTCMultiplayerPeer` nodes/classes mirror those names described there exactly. Mainly, because it's _just a wrapper_ for SIPSorcery. You will also need to come up with your own _Session Exchange Gateway_ (== What the Match Maker fundamentally is). This can be as simple as a text file on a remote server.

---

So what I am fundamentally failing to understand here is how testing the wrapper, and making changes to it, will help people understand the code. If you want to understand how WebRTC and such work, you should be toying around with SIPSorcery's source code. Changing my wrappers doesn't really help you understand anything.

That being said, like I said above, if you end up actually writing unit tests, possibly "real" integration tests, please open a PR, so I can review them :)
Though, keep in mind that they should truly test everything then. Including remote connections, special connection properties and especially if TURN servers work or not.
Said tests ideally should also work inside a CI (i.e. here on GitHub).

---

Let me know if there is anything else I can help you with ~
